File: docx-parser-converter-js/src/parsers/tables/table-row-parser.ts
// src/parsers/tables/table-row-parser.ts
import { TableRow, TableCell } from '../../models/table-models';
import { parseTableRowProperties } from './table-row-properties-parser';
import { parseTableCell } from './table-cell-parser';
import { getElement, getElements } from '../../utils/xml-utils';

export function parseTableRow(trNode: any): TableRow | null {
  if (!trNode || (typeof trNode !== 'object' && trNode !== true)) { // trNode can be true for <w:tr/>
    return null;
  }
  
  // Handle <w:tr/> which parses as `true` or an empty object for <w:tr></w:tr>
  if (trNode === true || (typeof trNode === 'object' && Object.keys(trNode).length === 0 && trNode.constructor === Object)) {
    return { type: 'tableRow', children: [], properties: undefined };
  }

  const trPrNode = getElement(trNode, 'w:trPr');
  const parsedTrProperties = trPrNode ? parseTableRowProperties(trPrNode) : undefined;

  const tcNodes = getElements(trNode, 'w:tc');
  const parsedCells: TableCell[] = [];
  for (const tcNode of tcNodes) {
    const cell = parseTableCell(tcNode);
    if (cell) {
      parsedCells.push(cell);
    }
  }

  // A row might technically be empty of cells in malformed XML,
  // but usually it should have at least one. The model allows empty children.
  return {
    type: 'tableRow',
    children: parsedCells,
    properties: parsedTrProperties,
  };
}

File: docx_parser_converter_python/parsers/document/table_row_parser.py
"""Parser for table row elements."""

from lxml.etree import _Element as Element

from models.document.table_row import TableRow
from parsers.document.table_cell_parser import parse_table_cell
from parsers.document.table_row_properties_parser import parse_table_row_properties
from parsers.utils import find_all_children, find_child


def parse_table_row(element: Element | None) -> TableRow | None:
    """Parse <w:tr> element.

    XML Example:
        <w:tr>
            <w:trPr>
                <w:tblHeader/>
            </w:trPr>
            <w:tc>...</w:tc>
            <w:tc>...</w:tc>
        </w:tr>

    Args:
        element: The <w:tr> element or None

    Returns:
        TableRow model or None if element is None
    """
    if element is None:
        return None

    # Parse row properties
    tr_pr = parse_table_row_properties(find_child(element, "trPr"))

    # Parse cells
    cells = []
    for cell_elem in find_all_children(element, "tc"):
        cell = parse_table_cell(cell_elem)
        if cell is not None:
            cells.append(cell)

    return TableRow(
        tr_pr=tr_pr,
        tc=cells,
    )
